// @polkadot/types/Metadata/v0/Events.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OuterEventMetadata = exports.OuterEventEventMetadata = exports.EventMetadata = void 0;

var _Struct = _interopRequireDefault(require("../../codec/Struct"));

var _Tuple = _interopRequireDefault(require("../../codec/Tuple"));

var _Vector = _interopRequireDefault(require("../../codec/Vector"));

var _Text = _interopRequireDefault(require("../../primitive/Text"));

var _Type = _interopRequireDefault(require("../../primitive/Type"));

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class EventMetadata extends _Struct.default {
  constructor(value) {
    super({
      name: _Text.default,
      args: _Vector.default.with(_Type.default),
      documentation: _Vector.default.with(_Text.default)
    }, value);
  }
  /**
   * @description The arguments of [[Type]]
   */


  get args() {
    return this.get('args');
  }
  /**
   * @description The arguments of [[Type]]
   * @deprecated Use `.args` instead
   */


  get arguments() {
    return this.args;
  }
  /**
   * @description The [[Text]] documentation
   */


  get documentation() {
    return this.get('documentation');
  }
  /**
   * @description The [[Text]] documentation
   * @deprecated Use `.documentation` instead.
   */


  get docs() {
    return this.documentation;
  }
  /**
   * @description The name for the event
   */


  get name() {
    return this.get('name');
  }

}

exports.EventMetadata = EventMetadata;

class OuterEventEventMetadata extends _Tuple.default {
  constructor(value) {
    super({
      Text: _Text.default,
      'Vec<EventMetadata>': _Vector.default.with(EventMetadata)
    }, value);
  }
  /**
   * @description The [[EventMetadata]]
   */


  get events() {
    return this[1];
  }
  /**
   * @description The name of the section
   */


  get name() {
    return this[0];
  }

}

exports.OuterEventEventMetadata = OuterEventEventMetadata;

class OuterEventMetadata extends _Struct.default {
  constructor(value) {
    super({
      name: _Text.default,
      events: _Vector.default.with(OuterEventEventMetadata)
    }, value);
  }
  /**
   * @description The [[OuterEventEventMetadata]]
   */


  get events() {
    return this.get('events');
  }
  /**
   * @description The name of the event
   */


  get name() {
    return this.get('name');
  }

}

exports.OuterEventMetadata = OuterEventMetadata;
