// @polkadot/types/primitive/Address.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ACCOUNT_ID_PREFIX = void 0;

var _util = require("@polkadot/util");

var _utilCrypto = require("@polkadot/util-crypto");

var _Base = _interopRequireDefault(require("../codec/Base"));

var _AccountId = _interopRequireDefault(require("./AccountId"));

var _AccountIndex = _interopRequireDefault(require("./AccountIndex"));

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
const ACCOUNT_ID_PREFIX = new Uint8Array([0xff]);
/**
 * @name Address
 * @description
 * A wrapper around an AccountId and/or AccountIndex that is encoded with a prefix.
 * Since we are dealing with underlying publicKeys (or shorter encoded addresses),
 * we extend from Base with an AccountId/AccountIndex wrapper. Basically the Address
 * is encoded as `[ <prefix-byte>, ...publicKey/...bytes ]` as per spec
 */

exports.ACCOUNT_ID_PREFIX = ACCOUNT_ID_PREFIX;

class Address extends _Base.default {
  constructor() {
    let value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : new Uint8Array();
    super(Address.decodeAddress(value));
  }

  static decodeAddress(value) {
    if (value instanceof _AccountId.default || value instanceof _AccountIndex.default) {
      return value;
    } else if ((0, _util.isBn)(value) || (0, _util.isNumber)(value)) {
      return new _AccountIndex.default(value);
    } else if (value instanceof Address) {
      return value.raw;
    } else if (Array.isArray(value)) {
      return Address.decodeAddress((0, _util.u8aToU8a)(value));
    } else if ((0, _util.isU8a)(value)) {
      // This allows us to instantiate an address with a raw publicKey. Do this first before
      // we checking the first byte, otherwise we may split an already-existent valid address
      if (value.length === 32) {
        return new _AccountId.default(value);
      } else if (value[0] === 0xff) {
        return new _AccountId.default(value.subarray(1));
      }

      const [offset, length] = _AccountIndex.default.readLength(value);

      return new _AccountIndex.default((0, _util.u8aToBn)(value.subarray(offset, offset + length), true));
    } else if ((0, _util.isHex)(value)) {
      return Address.decodeAddress((0, _util.hexToU8a)(value));
    }

    const decoded = (0, _utilCrypto.decodeAddress)(value);
    return decoded.length === 32 ? new _AccountId.default(decoded) : new _AccountIndex.default((0, _util.u8aToBn)(decoded, true));
  }
  /**
   * @description The length of the value when encoded as a Uint8Array
   */


  get encodedLength() {
    const rawLength = this.rawLength;
    return rawLength + ( // for 1 byte AccountIndexes, we are not adding a specific prefix
    rawLength > 1 ? 1 : 0);
  }
  /**
   * @description Checks if the value is an empty value
   */


  get isEmpty() {
    return this.raw.isEmpty;
  }
  /**
   * @description The length of the raw value, either AccountIndex or AccountId
   */


  get rawLength() {
    return this.raw instanceof _AccountIndex.default ? _AccountIndex.default.calcLength(this.raw) : this.raw.encodedLength;
  }
  /**
   * @description Compares the value of the input to see if there is a match
   */


  eq(other) {
    return this.raw.eq(other);
  }
  /**
   * @description Returns a hex string representation of the value
   */


  toHex() {
    return (0, _util.u8aToHex)(this.toU8a());
  }
  /**
   * @description Converts the Object to JSON, typically used for RPC transfers
   */


  toJSON() {
    return this.raw.toJSON();
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return this.raw.toString();
  }
  /**
   * @description Returns the base runtime type name for this instance
   */


  toRawType() {
    return 'Address';
  }
  /**
   * @description Encodes the value as a Uint8Array as per the SCALE specifications
   * @param isBare true when the value has none of the type-specific prefixes (internal)
   */


  toU8a(isBare) {
    const encoded = this.raw.toU8a().subarray(0, this.rawLength);
    return isBare ? encoded : (0, _util.u8aConcat)(this.raw instanceof _AccountIndex.default ? _AccountIndex.default.writeLength(encoded) : ACCOUNT_ID_PREFIX, encoded);
  }

}

exports.default = Address;
