// @polkadot/types/type/PrefabWasmModule.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Compact = _interopRequireDefault(require("../codec/Compact"));

var _Option = _interopRequireDefault(require("../codec/Option"));

var _Struct = _interopRequireDefault(require("../codec/Struct"));

var _Bytes = _interopRequireDefault(require("../primitive/Bytes"));

var _U = _interopRequireDefault(require("../primitive/U32"));

// Copyright 2017-2019 @polkadot/types authors & contributors
// This software may be modified and distributed under the terms
// of the Apache-2.0 license. See the LICENSE file for details.
class PrefabWasmModuleReserved extends _Option.default {}
/**
 * @name PrefabWasmModule
 * @description Struct to encode the vesting schedule of an individual account
 */


class PrefabWasmModule extends _Struct.default {
  constructor(value) {
    super({
      scheduleVersion: _Compact.default.with(_U.default),
      initial: _Compact.default.with(_U.default),
      maximum: _Compact.default.with(_U.default),
      _reserved: PrefabWasmModuleReserved,
      code: _Bytes.default
    }, value);
  }
  /**
   * @description The code as [[Bytes]]
   */


  get code() {
    return this.get('code');
  }
  /**
   * @description The initial as [[Compact]]
   */


  get initial() {
    return this.get('initial');
  }
  /**
   * @description The maximum as [[Compact]]
   */


  get maximum() {
    return this.get('maximum');
  }
  /**
   * @description The scheduleVersion value as [[Compact]]
   */


  get scheduleVersion() {
    return this.get('scheduleVersion');
  }

}

exports.default = PrefabWasmModule;
